src/lib/sticky-header/sticky-header-dir.ts

```diff
@@ -0,0 +1,399 @@
+/** Class applied when the header is not "stuck" */
+const STICK_END_CLASS = 'cdk-sticky-header-end';
+/**
+ * Set a debounce time which is used in debounce() function when adding event listeners.
+ * Set is as 5. Because if the DEBOUNCE_TIME is set as a too large number. The sticky effect
+ * during scroll will become vary strange and can not be scrolled smoothly.
+ */
+const DEBOUNCE_TIME: number = 5;
+/**
+ * Directive that marks an element as a sticky-header. Inside of a scrolling container (marked with
+ * cdkScrollable), this header will "stick" to the top of the scrolling viewport while its sticky
+ * region (see cdkStickyRegion) is in view.
+ */
+@Directive({
+  selector: '[cdkStickyHeader]',
+})
+export class CdkStickyHeader implements OnDestroy, AfterViewInit {
+
+  /**
+   * Set the sticky-header's z-index as 10 in default. Make it as an input
+   * variable to make user be able to customize the zIndex when
+   * the sticky-header's zIndex is not the largest in current page.
+   * Because if the sticky-header's zIndex is not the largest in current page,
+   * it may be sheltered by other element when being stuck.
+   */
+  @Input('cdkStickyHeaderZIndex') zIndex: number = 10;
+
+  isStuck: boolean = false;
+  /** Whether the browser support CSS sticky positioning. */
+  isStickyPositionSupported: boolean = true;
+
+
+  /** The element with the 'cdkStickyHeader' tag. */
+  element: HTMLElement;
+  /** The upper container element with the 'cdkStickyRegion' tag. */
+  stickyParent: HTMLElement | null;
+  /** The upper scrollable container. */
+  upperScrollableContainer: HTMLElement;
+  /**
+   * The original css of the sticky element, used to reset the sticky element
+   * when it is being unstuck
+   */
+  originalCss: any;
+  /**
+   * 'getBoundingClientRect().top' of CdkStickyRegion of current sticky header.
+   * It is used with '_scrollFinish' to judge whether the current header
+   * need to be stuck.
+   */
+  private _containerStart: number;
+  /**
+   * `_scrollFinish` is the place from where the stuck element should be unstuck
+   */
+  private _scrollFinish: number;
+
+  private _onScrollSubscription: Subscription;
+
+  private _onTouchSubscription: Subscription;
+
+  private _onResizeSubscription: Subscription;
+
+  constructor(element: ElementRef,
+              scrollable: Scrollable,
+              @Optional() public parentRegion: CdkStickyRegion,
+              platform: Platform) {
+    if (platform.isBrowser) {
+      this.element = element.nativeElement;
+      this.upperScrollableContainer = scrollable.getElementRef().nativeElement;
+      this.setStrategyAccordingToCompatibility();
+    }
+  }
+
+  ngAfterViewInit(): void {
+    if (!this.isStickyPositionSupported) {
+      this.stickyParent = this.parentRegion != null ?
+        this.parentRegion._elementRef.nativeElement : this.element.parentElement;
+      let values = window.getComputedStyle(this.element, '');
+      this.originalCss = this.generateCssStyle(
```

Comment:
The `generateCssStyle` isn't necessary; it doesn't shorten the code or capture any behavior. It's simpler to just use an object literal here.

The `getPropertyValue` method shouldn't be necessary for the result of `getComputedStyle`; you should be able to just access the property directly, e.g., `computedStyle.display`

If you want to capture the original styles and re-apply them when you're done, you don't want to save the results of `getComputedStyle` because these styles are _computed_. To restore things as they were, you would need to save the values from the `style` property of the element. If you need any of the computed styles for other reasons, then you shoudl store those separately from the `originalCss` that's used for restoration.